### DISCUSSION TITLE
hash collision

### DISCUSSION THREAD

**Discussion by hustchaya** [async-profiler/async-profiler#931] [Q&A] [answered]
do we need to consider hash collision in class_map/call_trace_storage? why?

**Comment by apangin** [comment]
Could you elaborate on the problem you are seeing with hashing?

**Reply by hustchaya** [reply]
No problems encountered in practice, but collisions may exist in theory ? Could you tell me your considerations? Is it ignored because the probability of collision is very small? Would you happen to have any experimental data available?

**Reply by apangin** [reply]
I'm not sure what conflicts you are referring to. Both `_class_map` and `_call_trace_storage` are exact hash maps, meaning they store all keys, properly handling possible hash collisions. The only assumption I make is that we don't need to compare whole stack traces byte-by-byte, instead, it's enough to compare their 64-bit MurmurHash fingerprints.
For the reference, the probability of collision among 1 million of these fingerprints is about ~0.000000027.

**Reply by hustchaya** [reply]
Sorry, I didn’t translate the professional terms clearly. What I want to express is if there is a hash collision of 0.000000027, will there be any problems with the program logic?The hash algorithm of _class_map has a higher probability of hash collision, right?Should we consider this low probability scenario?

**Reply by apangin** [reply]
As mentioned above, `_class_map` and `_call_trace_storage` already handle collisions, I don't see a problem there.

**Reply by hustchaya** [reply]
![image](https://github.com/async-profiler/async-profiler/assets/102569465/13edaefb-f346-4b2a-a536-545380309b4b)
At the location of String -> hash, is it possible that there are two different Strings that are calculated to be the same hash?

**Reply by apangin** [reply, marked as answer]
`LongHashTables` stores 64-bit integer keys, not strings. `u64 hash` argument should be actually called `u64 key` - this is what might have caused confusion.

For simplicity and performance of the implementation, async-profiler reduces stack traces to 64-bit keys with the collision probability of about 2.7*10<sup>-8</sup> for 1M elements, meaning that you'll never see it in practice (even if this could happen, it would not affect overall profile quality).
